src/lib/operators/aggregate_hash.cpp

```diff
@@ -698,7 +785,9 @@ std::shared_ptr<const Table> AggregateHash::_on_execute() {
     auto pos_list = RowIDPosList();
     pos_list.reserve(context->results.size());
     for (const auto& result : context->results) {
-      pos_list.push_back(result.row_id);
+      // NULL means that this result was overallocated (see get_or_add_result) and must be ignored
```

Comment:
Can you tell me where to  look at exactly? I can't find the overallocation part.

Reply:
> If we have not seen this index as part of the current aggregate function, the results vector may not yet have the correct size. Resize it if necessary and write the current row_id so that we can recover the GroupBy column(s) later. By default, the newly created values have an invalid RowID and are later ignored. We grow the vector slightly more than necessary. Otherwise, monotonically increasing keys would lead to one resize per row.

Reply:
Why does NULL mean overallocated? Could not just be a value gap in the groupby keys?

Reply:
No, see aggregate_hash.hpp:

> There is no ambiguity with actual NULLs because the aggregate operator is not NULL-producing. As such, we know that each valid GROUP BY-group has at least one valid input RowID. As the input may be a ReferenceSegment, a valid RowID may *point to* a row that is NULL.

I am happy to add any explanation that makes this clearer. I am not particularly happy with it myself.

Reply:
I thought a shortcut aggregation where do not have all the keys within min/max actually exist, would lead to "gaps" where we have a NULL_ROW_ID and thus a NULL_ROW_ID would not mean an overallocation. 

In the case of the offset caching, we need to resize and thus I understand the overallocation part. But for the shortcut, it's "just a gap" isn't it?

Reply:
Yes. Would you like 

> NULL means that this result is either a gap (in the case of an unused immediate key) or the result of overallocating the result vector.

better?

Reply:
Jep. 👍